src/mc-fle2-payload-iev-v2.c

```diff
@@ -0,0 +1,388 @@
+   bool parsed;
+   bool ClientEncryptedValueDecoded;
+   bool ClientValueDecoded;
+
+   // Populated during _add_S_Key
+   // DecryptedServerEncryptedValue := DecryptCTR(S_Key, ServerEncryptedValue)
+   _mongocrypt_buffer_t DecryptedServerEncryptedValue;
+
+   // Views on DecryptedServerEncryptedValue (DSEV)
+   _mongocrypt_buffer_t K_KeyId;              // First 16 octets, UUID
+   _mongocrypt_buffer_t ClientEncryptedValue; // Remainder of DSEV
+
+   // Populated during _add_K_Key
+   // ClientValue := DecryptCBCAEAD(K_Key, ClientEncryptedValue, AD=K_KeyId)
+   _mongocrypt_buffer_t ClientValue;
+};
+
+#define kMetadataLen 96U                // encCount(32) + tag(32) + encZeros(32)
+#define kMinServerEncryptedValueLen 17U // IV(16) + EncrtyptCTR(1byte)
+#define kMinSEVAndMetadataLen (kMinServerEncryptedValueLen + kMetadataLen)
+
+#define CHECK_AND_RETURN(x) \
+   if (!(x)) {              \
+      return false;         \
+   }
+
+mc_FLE2IndexedEqualityEncryptedValueV2_t *
+mc_FLE2IndexedEqualityEncryptedValueV2_new (void)
+{
+   return bson_malloc0 (sizeof (mc_FLE2IndexedEqualityEncryptedValueV2_t));
+}
+
+bool
+mc_FLE2IndexedEqualityEncryptedValueV2_parse (
+   mc_FLE2IndexedEqualityEncryptedValueV2_t *iev,
+   const _mongocrypt_buffer_t *buf,
+   mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (iev);
+   BSON_ASSERT_PARAM (buf);
+
+   if (iev->parsed) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValueV2_parse must not be "
+                  "called twice");
+      return false;
+   }
+
+   mc_reader_t reader;
+   mc_reader_init_from_buffer (&reader, buf, __FUNCTION__);
+
+   CHECK_AND_RETURN (
+      mc_reader_read_u8 (&reader, &iev->fle_blob_subtype, status));
+
+   if (iev->fle_blob_subtype !=
+       MC_SUBTYPE_FLE2IndexedEqualityEncryptedValueV2) {
+      CLIENT_ERR ("mc_FLE2IndexedEqualityEncryptedValueV2_parse expected "
+                  "fle_blob_subtype %d got: %" PRIu8,
+                  MC_SUBTYPE_FLE2IndexedEqualityEncryptedValueV2,
+                  iev->fle_blob_subtype);
+      return false;
+   }
+
+   /* Read S_KeyId. */
+   CHECK_AND_RETURN (
+      mc_reader_read_uuid_buffer (&reader, &iev->S_KeyId, status));
+
+   /* Read original_bson_type. */
+   CHECK_AND_RETURN (
+      mc_reader_read_u8 (&reader, &iev->bson_value_type, status));
+
+   /* Read ServerEncryptedValue. */
+   const uint64_t SEV_and_metadata_len =
+      mc_reader_get_remaining_length (&reader);
+   if (SEV_and_metadata_len < kMinSEVAndMetadataLen) {
+      CLIENT_ERR ("Invalid payload size %" PRIu64
+                  ", smaller than minimum length %d",
+                  SEV_and_metadata_len,
+                  kMinSEVAndMetadataLen);
+      return false;
+   }
+   const uint64_t SEV_len = SEV_and_metadata_len - kMetadataLen;
+   CHECK_AND_RETURN (mc_reader_read_buffer (
+      &reader, &iev->ServerEncryptedValue, SEV_len, status));
+
+   // Ignore Metadata block.
+   BSON_ASSERT (mc_reader_get_remaining_length (&reader) == kMetadataLen);
```

Comment:
The reason that it asserts rather than erroring is because we've explicitly calculated out position to be at this location when we created `SEV_len`.  If the remaining length is even NOT `kMetadataLen` at this point then the reader implementation has failed in some spectacular way.